### ISSUE TITLE
Reason for the discrete distribution

### ISSUE THREAD

**Issue by dgm2** [Liuhong99/Sophia#49]
From the code, Sophia does the following

- the default step: 
  - line 5 from Algorithm 1 from Sophia paper
  - gets gradients from training set
- the hessian step (every k iterations)
  - part1:
    - gets the logits and loss from training set
    - computes the gradients
  - part2:
    - gets logits from training set without label
    - samples "labels" from the logits with discrete distribution (line 4 in Algorithm 2 from Sophia paper)
    - compute loss between "labels" and logits
    - update hessian

I could not find a reason for using such discrete distribution over the prediction. This is the CE between the samples over the prediction and the prediction. Could you provide some intuition about the need of such distribution, please. Why not getting it from training data? Are there some computational advantages or something more basic that I am missing?

Thanks for your work! 
cc @Liuhong99 @tengyuma

**Comment by dgm2** [comment]
@Liuhong99 In the paper I could not find the explanation of why there are the two following types of gradients.

1. The first is from Line  5 Alg 3 (Sophia) CE between predictions and labels. ( this is easy to understand)
2. The second Line 4 Alg 2 (GNB estimator). Specifically, in the GNB gradients are computed via a CE between a sample over predictions, and the predictions. This one I cannot understand. Why this type of gradient is used here?

Thanks & Best regards

**Comment by Liuhong99** [comment]
Thanks for the interest in the paper! The resampling of labels is to make the estimate of the Gauss-Newton diagonal terms unbiased. The derivation is in page 8 of our paper (https://arxiv.org/abs/2305.14342). As also noted by Kunstner et al, if the labels are from data distribution, then the estimate is no longer unbiased (termed empirical fisher).  

Kunstner, F., Hennig, P., and Balles, L. Limitations of the empirical fisher approximation for natural gradient descent
